Graphs/js/main-directed.js

```diff
@@ -827,3 +827,143 @@ function myPageRank() {
         animateNodes(ranks, 1000);
     }
 }
+
+function heldKarp() {
+    stopAnimation();
+    let root;
+    const [wm, id_index] = getWM(cy, false, true);
+
+    if(!isComplete()){
+        alert('Graph not complete. Please set the probability that two nodes are connected to 1.');
+        return;
+    } 
+
+    if (cy.elements(':selected').length >= 1) {
+        root = cy.elements(':selected')[0];
+    } else {
+        root = prompt('Please enter the id of the start node.\n', '1');
+        root = parseInt(root) - 1;
+    }
+
+    clearCyStyle();
+    
+    const nodes = getAllNodes(cy);
+
+    cy.elements(':grabbable').unselect();
+
+    heldKarpHelp(wm, root);
+}
+
+/**
+ *
+ * @param {number[][]} wm weight matrix
+ * @param {number} x starting point
+ */
+function heldKarpHelp(wm, x) {
+    let n = wm.length;
+    var memo = new Map();
+
+    let nums = new Array(n);
+    for (let i = 0; i < n; i++) {
+        nums[i] = i;
+    }
+    nums.splice(x, 1);
+
+    let rst = new Array(n - 1);
+    let min = Infinity;
+    let minArr;
+
+    for (let i = 0; i < nums.length; i++) {
+        let tempNums = nums.concat();
+        tempNums.splice(i, 1);
+        let h = heldKarpPath(nums[i], x, tempNums, wm, memo);
+        rst[i] = wm[x][nums[i]] + h.distance;
+
+        if (wm[x][nums[i]] + h.distance < min) {
+            min = wm[x][nums[i]] + h.distance;
+            minArr = h.arr.concat();
+        }
+    }
+
+    let minArrInv = [];
+
+    for (let i = 0; i < minArr.length; i++) {
+        minArrInv.push(minArr[minArr.length - i - 1]);
+    }
+
+    minArrInv.unshift(x);
+
+    let pathList = new LinkedList();
+    let path = new Array();
+    let nodes = getAllNodes(cy);
+
+    for (let i = 0; i < minArrInv.length; i++) {
+        path.push(nodes[minArrInv[i]]);
+        pathList.add(nodes[minArrInv[i]]);
+        path.push(nodes[minArrInv[i]].edgesTo(nodes[minArrInv[(i + 1) % minArrInv.length]])[0]);
+        pathList.add(nodes[minArrInv[i]].edgesTo(nodes[minArrInv[(i + 1) % minArrInv.length]])[0]);
+    }
+    
+    clearCyStyle();
+
+    let val = "";
+
+    for(let i = 0; i < minArrInv.length; i++){
+        val += (minArrInv[i] + 1) + "→"
+    }
+
+    val += (x + 1) + "</br>";
+    val += "Sum of weight of the shortest path is " + min;
+    processDiv.innerHTML = val;
+
+    pathList.traverse(animation_check.checked, true);
+
+    return { distance: min, arr: minArr };
+}
+
+/**
+ *
+ * @param {number} x current node
+ * @param {number} rt starting node
+ * @param {number[]} nums nodes to connect
+ * @param {number[][]} wm distance matrix; wm(ij) represents distance from i to j
+ * @param {Map} memo
```

Comment:
You can annotate Map objects like this
```suggestion
 * @param {Map<string, {distance: number, arr: number[]}>} memo
```